swift_browser_ui_frontend/src/common/download.js

```diff
@@ -0,0 +1,230 @@
+          ),
+        );
+        this.chunk = undefined;
+      }
+
+      if (this.chunkBuffer.length > 0) {
+        navigator.serviceWorker.ready.then(reg => {
+          reg.active.postMessage({
+            cmd: "decryptChunk",
+            chunk: new Uint8Array(this.chunkBuffer),
+          });
+          this.chunkBuffer = [];
+        });
+      } else {
+        console.log("Telling the serviceWorker that decryption is done.");
+        navigator.serviceWorker.ready.then(reg => {
+          reg.active.postMessage({
+            cmd: "decryptionFinished",
+          });
+        });
+      }
+    }
+  }
+
+  getFileUrl() {
+    let fileURL = new URL(
+      `/download/${this.project}/${this.container}/${this.object}`,
+      document.location.origin,
+    );
+    fileURL.searchParams.append("project", this.active.id);
+    return fileURL;
+  }
+
+  async getHeader(pubkey) {
+    let upInfo = await getUploadEndpoint(
+      this.active.id,
+      this.project,
+      this.container,
+    );
+    this.$store.commit("setUploadInfo", upInfo);
+
+    let signatureUrl = new URL(`/sign/${60}`, document.location.origin);
+    signatureUrl.searchParams.append("path", this.whitelistPath);
+    let signed = await GET(signatureUrl);
+    signed = await signed.json();
+    let whitelistUrl = new URL(this.whitelistPath, this.endpoint);
+    whitelistUrl.searchParams.append("valid", signed.valid);
+    whitelistUrl.searchParams.append("signature", signed.signature);
+    whitelistUrl.searchParams.append("flavor", "crypt4gh");
+    whitelistUrl.searchParams.append(
+      "session",
+      this.$store.state.uploadInfo.id,
+    );
+    let resp = await PUT(whitelistUrl, pubkey);
+    let headerPath = `/header/${this.active.id}/${this.container}/${this.object}`;
+    signatureUrl = new URL(`/sign/${60}`, document.location.origin);
+    signatureUrl.searchParams.append("path", headerPath);
+    signed = await GET(signatureUrl);
+    signed = await signed.json();
+    let headerUrl = new URL(this.endpoint.concat(headerPath));
+    headerUrl.searchParams.append("valid", signed.valid);
+    headerUrl.searchParams.append("signature", signed.signature);
+    headerUrl.searchParams.append(
+      "session",
+      this.$store.state.uploadInfo.id,
+    );
+    resp = await GET(headerUrl);
+    let header = await resp.text();
+    header = Uint8Array.from(atob(header), c => c.charCodeAt(0));
+    addHeader(
+      header,
+      this.object.split("/").pop(),
+      undefined,
+    );
+  }
+
+  initServiceWorker() {
+    navigator.serviceWorker.addEventListener("message", (e) => {
+      e.stopImmediatePropagation();
+      switch (e.data.eventType) {
+        case "downloadSessionOpened":
+          this.object = this.objects.pop();
+          this.currentFinished = false;
+          if (this.object == undefined) {
+            this.finished = true;
+          }
+          this.getHeader(e.data.pubKey).then(() => { });
```

Comment:
Why not use `await`?

Reply:
I've occasionally encountered problems with async functions as event handlers, in cases with a congested event loop. If I move to asynchronous event handler function I'd need to add some ordering number to the downloaded chunks, and alter the slicer a bit. This way I can be sure that the slicer calls are fired in the correct order, and chunks arrive in the correct order to the `ServiceWorker` for decryption.

I have some future improvements in mind, which will make the implementation overall more robust and parallel, but those don't really make sense to add into this initial implementation.